File: 29 Divide Two Integers/29 LL.py
class Solution(object):
	def divide(self, dividend, divisor):
		neg = (dividend < 0) ^ (divisor < 0)
		dividend = abs(dividend)
		divisor = abs(divisor)
		lshift = 0
		mask = divisor
		mask |= mask >> 1
		mask |= mask >> 2
		mask |= mask >> 4
		mask |= mask >> 8
		mask |= mask >> 16
		while dividend > mask:
			lshift += 1
			divisor <<= 1
			mask = (mask << 1) | 1
		mask = 0
		while lshift >= 0:
			if dividend >= divisor:
				dividend -= divisor
				mask |= (1 << lshift)
			divisor >>= 1
			lshift -= 1
		if not neg and mask > 0x7FFFFFFF:
			return 0x7FFFFFFF
		return -mask if neg else mask

File: 29 Divide Two Integers/29 LL.cpp
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/divide-two-integers/

Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.
*/

class Solution
{
public:
	int divide(int dividend, int divisor)
	{
		if (divisor == 0) return 0x7FFFFFFF;
		bool neg = (dividend < 0) ^ (divisor < 0);
		unsigned int a = (dividend < 0) ? -dividend : dividend;
		unsigned int b = (divisor < 0) ? -divisor : divisor;
		unsigned int result = 0, mask = 1;
		while ((a >> 1) >= b)
		{
			b <<= 1;
			mask <<= 1;
		}
		while (mask != 0)
		{
			if (a >= b)
			{
				a -= b;
				result |= mask;
			}
			b >>= 1;
			mask >>= 1;
		}
		// attention unsigned to signed MIN_INT overflow
		return neg ? -int(result) : int((result > 0x7FFFFFFF) ? 0x7FFFFFFF : result);
	}
};

/*
idea: bit subtraction and shift
complexity: Time O(N)
*/

int main(void)
{
	Solution engine;
	cout << engine.divide(0x80000001, 1) << '\n';
	return 0;
}